store/views.py

```diff
@@ -1,82 +1,136 @@
+        book=book, username=user)
+    change_rating.rating = rating
+    change_rating.save()
+
+    book.rating = BookRating.objects.filter(
+        book=book).aggregate(rating=Avg('rating'))['rating']
+    book.save()
+    
+    return redirect("book-detail", bid=bid)
+
+
 
 @csrf_exempt
 def bookListView(request):
     template_name = 'store/book_list.html'
     context = {
-        'books': None, # set this to the list of required books upon filtering using the GET parameters
-                       # (i.e. the book search feature will also be implemented in this view)
+        'books': None,
     }
     get_data = request.GET
-    # START YOUR CODE HERE
-    
+
+    if get_data:
+        context['books'] = Book.objects.filter(
+            title__icontains=get_data['title'], author__icontains=get_data['author'], genre__icontains=get_data['genre']
+        )
+        
+    else:
+        context['books'] = Book.objects.all()
     
+    for i in context['books']:
+        i.rating = BookRating.objects.filter(
+        book=i).aggregate(rating=Avg('rating'))['rating']
+
     return render(request, template_name, context=context)
 
+
 @login_required
 def viewLoanedBooks(request):
     template_name = 'store/loaned_books.html'
     context = {
         'books': None,
     }
-    '''
-    The above key 'books' in the context dictionary should contain a list of instances of the 
-    BookCopy model. Only those book copies should be included which have been loaned by the user.
-    '''
-    # START YOUR CODE HERE
-    
-
-
+    user = request.user
+    books = BookCopy.objects.filter(borrower=user)
+    context['books'] = books
     return render(request, template_name, context=context)
 
+
 @csrf_exempt
 @login_required
 def loanBookView(request):
     response_data = {
         'message': None,
     }
-    '''
-    Check if an instance of the asked book is available.
-    If yes, then set the message to 'success', otherwise 'failure'
-    '''
-    # START YOUR CODE HERE
-    book_id = None # get the book id from post data
-
+    bid = request.POST['bid']
+    book = BookCopy.objects.filter(book_id=bid, status=True)[0]
 
+    if book:
+        book.borrower = request.user
+        book.borrow_date = timezone.datetime.today().date()
+        book.status = False
+        book.save()
+        response_data['message'] = 'success'
+    else:
+        response_data['message'] = 'failure'
     return JsonResponse(response_data)
 
-'''
-FILL IN THE BELOW VIEW BY YOURSELF.
-This view will return the issued book.
-You need to accept the book id as argument from a post request.
-You additionally need to complete the returnBook function in the loaned_books.html file
-to make this feature complete
-''' 
+
 @csrf_exempt
 @login_required
 def returnBookView(request):
-    pass
+    response_data = {
+        'message': None,
+    }
 
+    bid = request.POST['bid']
+    book = get_object_or_404(BookCopy, pk=bid)
 
+    if book:
+        book.borrower = None
```

Comment:
During the book return, you must validate that the borrower of the book is the current user.